### PR TITLE
[POC] feat(ui-react-native): Add react-native theme output

### PR DESCRIPTION
**Pull request by ioanabrooks**
#### Description of changes

This change adds a React Native theme output.

Credit: heavily inspired by https://github.com/aws-amplify/amplify-ui/pull/157

#### Description of how you validated changes

Output of `yarn ui build`:

```
/**
 * Do not edit directly
 * Generated on Wed, 14 Sep 2022 17:47:19 GMT
 */

const theme = {
  "borderWidths": {
    "small": 1,
    "medium": 2,
    "large": 3
  },
  "colors": {
    "red": {
      "10": "hsl(0, 75%, 95%)",
      "20": "hsl(0, 75%, 85%)",
      "40": "hsl(0, 75%, 75%)",
      "60": "hsl(0, 50%, 50%)",
      "80": "hsl(0, 95%, 30%)",
      "90": "hsl(0, 100%, 20%)",
      "100": "hsl(0, 100%, 15%)"
    },
    "orange": {
      "10": "hsl(30, 75%, 95%)",
      "20": "hsl(30, 75%, 85%)",
      "40": "hsl(30, 75%, 75%)",
      "60": "hsl(30, 50%, 50%)",
      "80": "hsl(30, 95%, 30%)",
      "90": "hsl(30, 100%, 20%)",
      "100": "hsl(30, 100%, 15%)"
    },
    "yellow": {
      "10": "hsl(60, 75%, 95%)",
      "20": "hsl(60, 75%, 85%)",
      "40": "hsl(60, 75%, 75%)",
      "60": "hsl(60, 50%, 50%)",
      "80": "hsl(60, 95%, 30%)",
      "90": "hsl(60, 100%, 20%)",
      "100": "hsl(60, 100%, 15%)"
    },
    "green": {
      "10": "hsl(130, 60%, 95%)",
      "20": "hsl(130, 60%, 90%)",
      "40": "hsl(130, 44%, 63%)",
      "60": "hsl(130, 43%, 46%)",
      "80": "hsl(130, 33%, 37%)",
      "90": "hsl(130, 27%, 29%)",
      "100": "hsl(130, 22%, 23%)"
    },
    "teal": {
      "10": "hsl(190, 75%, 95%)",
      "20": "hsl(190, 75%, 85%)",
      "40": "hsl(190, 70%, 70%)",
      "60": "hsl(190, 50%, 50%)",
      "80": "hsl(190, 95%, 30%)",
      "90": "hsl(190, 100%, 20%)",
      "100": "hsl(190, 100%, 15%)"
    },
    "blue": {
      "10": "hsl(220, 95%, 95%)",
      "20": "hsl(220, 85%, 85%)",
      "40": "hsl(220, 70%, 70%)",
      "60": "hsl(220, 50%, 50%)",
      "80": "hsl(220, 95%, 30%)",
      "90": "hsl(220, 100%, 20%)",
      "100": "hsl(220, 100%, 15%)"
    },
    "purple": {
      "10": "hsl(300, 95%, 95%)",
      "20": "hsl(300, 85%, 85%)",
      "40": "hsl(300, 70%, 70%)",
      "60": "hsl(300, 50%, 50%)",
      "80": "hsl(300, 95%, 30%)",
      "90": "hsl(300, 100%, 20%)",
      "100": "hsl(300, 100%, 15%)"
    },
    "pink": {
      "10": "hsl(340, 95%, 95%)",
      "20": "hsl(340, 90%, 85%)",
      "40": "hsl(340, 70%, 70%)",
      "60": "hsl(340, 50%, 50%)",
      "80": "hsl(340, 95%, 30%)",
      "90": "hsl(340, 100%, 20%)",
      "100": "hsl(340, 100%, 15%)"
    },
    "neutral": {
      "10": "hsl(210, 5%, 98%)",
      "20": "hsl(210, 5%, 94%)",
      "40": "hsl(210, 5%, 87%)",
      "60": "hsl(210, 8%, 55%)",
      "80": "hsl(210, 10%, 40%)",
      "90": "hsl(210, 25%, 25%)",
      "100": "hsl(210, 50%, 10%)"
    },
    "brand": {
      "primary": {
        "10": "hsl(190, 75%, 95%)",
        "20": "hsl(190, 75%, 85%)",
        "40": "hsl(190, 70%, 70%)",
        "60": "hsl(190, 50%, 50%)",
        "80": "hsl(190, 95%, 30%)",
        "90": "hsl(190, 100%, 20%)",
        "100": "hsl(190, 100%, 15%)"
      },
      "secondary": {
        "10": "hsl(300, 95%, 95%)",
        "20": "hsl(300, 85%, 85%)",
        "40": "hsl(300, 70%, 70%)",
        "60": "hsl(300, 50%, 50%)",
        "80": "hsl(300, 95%, 30%)",
        "90": "hsl(300, 100%, 20%)",
        "100": "hsl(300, 100%, 15%)"
      }
    },
    "font": {
      "primary": "hsl(210, 50%, 10%)",
      "secondary": "hsl(210, 25%, 25%)",
      "tertiary": "hsl(210, 10%, 40%)",
      "disabled": "hsl(210, 8%, 55%)",
      "inverse": "hsl(0, 0%, 100%)",
      "interactive": "hsl(190, 95%, 30%)",
      "active": "hsl(190, 100%, 15%)",
      "info": "hsl(220, 100%, 20%)",
      "warning": "hsl(30, 100%, 20%)",
      "error": "hsl(0, 100%, 20%)",
      "success": "hsl(130, 27%, 29%)"
    },
    "background": {
      "primary": "hsl(0, 0%, 100%)",
      "secondary": "hsl(210, 5%, 98%)",
      "tertiary": "hsl(210, 5%, 94%)",
      "quaternary": "hsl(210, 8%, 55%)",
      "disabled": "hsl(210, 5%, 94%)",
      "info": "hsl(220, 85%, 85%)",
      "warning": "hsl(30, 75%, 85%)",
      "error": "hsl(0, 75%, 85%)",
      "success": "hsl(130, 60%, 90%)"
    },
    "border": {
      "primary": "hsl(210, 8%, 55%)",
      "secondary": "hsl(210, 5%, 87%)",
      "tertiary": "hsl(210, 5%, 94%)",
      "disabled": "hsl(210, 5%, 94%)",
      "pressed": "hsl(190, 100%, 15%)",
      "error": "hsl(0, 95%, 30%)"
    },
    "shadow": {
      "primary": "hsla(210, 50%, 10%, 0.25)",
      "secondary": "hsla(210, 50%, 10%, 0.15)",
      "tertiary": "hsla(210, 50%, 10%, 0.05)"
    },
    "overlay": {
      "10": "hsla(0, 0%, 0%, 0.1)",
      "20": "hsla(0, 0%, 0%, 0.2)",
      "30": "hsla(0, 0%, 0%, 0.3)",
      "40": "hsla(0, 0%, 0%, 0.4)",
      "50": "hsla(0, 0%, 0%, 0.5)",
      "60": "hsla(0, 0%, 0%, 0.6)",
      "70": "hsla(0, 0%, 0%, 0.7)",
      "80": "hsla(0, 0%, 0%, 0.8)",
      "90": "hsla(0, 0%, 0%, 0.9)"
    },
    "black": "hsl(0, 0%, 0%)",
    "white": "hsl(0, 0%, 100%)",
    "transparent": "transparent"
  },
  "fonts": {
    "default": {
      "variable": "'InterVariable', 'Inter var', 'Inter', -apple-system, BlinkMacSystemFont,\n        'Helvetica Neue', 'Segoe UI', Oxygen, Ubuntu, Cantarell, 'Open Sans',\n        sans-serif",
      "static": "'Inter', -apple-system, BlinkMacSystemFont, 'Helvetica Neue',\n        'Segoe UI', Oxygen, Ubuntu, Cantarell, 'Open Sans', sans-serif"
    }
  },
  "fontSizes": {
    "xxxs": 6,
    "xxs": 8,
    "xs": 12,
    "small": 14,
    "medium": 16,
    "large": 20,
    "xl": 24,
    "xxl": 32,
    "xxxl": 40,
    "xxxxl": 48
  },
  "fontWeights": {
    "hairline": "100",
    "thin": "200",
    "light": "300",
    "normal": "400",
    "medium": "500",
    "semibold": "600",
    "bold": "700",
    "extrabold": "800",
    "black": "900"
  },
  "lineHeights": {
    "small": "1.25",
    "medium": "1.5",
    "large": "2"
  },
  "opacities": {
    "0": 0,
    "10": 0.1,
    "20": 0.2,
    "30": 0.3,
    "40": 0.4,
    "50": 0.5,
    "60": 0.6,
    "70": 0.7,
    "80": 0.8,
    "90": 0.9,
    "100": 1
  },
  "outlineOffsets": {
    "small": "1px",
    "medium": "2px",
    "large": "3px"
  },
  "outlineWidths": {
    "small": "1px",
    "medium": "2px",
    "large": "3px"
  },
  "radii": {
    "xs": 2,
    "small": 4,
    "medium": 8,
    "large": 16,
    "xl": 32,
    "xxl": 64,
    "xxxl": 128
  },
  "shadows": {
    "small": {
      "shadowColor": "hsla(210, 50%, 10%, 0.05)",
      "shadowOffset": {
        "width": 0,
        "height": 2
      },
      "shadowRadius": 4
    },
    "medium": {
      "shadowColor": "hsla(210, 50%, 10%, 0.15)",
      "shadowOffset": {
        "width": 0,
        "height": 2
      },
      "shadowRadius": 6
    },
    "large": {
      "shadowColor": "hsla(210, 50%, 10%, 0.25)",
      "shadowOffset": {
        "width": 0,
        "height": 4
      },
      "shadowRadius": 12
    }
  },
  "space": {
    "zero": 0,
    "xxxs": 2,
    "xxs": 3,
    "xs": 4,
    "small": 6,
    "medium": 8,
    "large": 12,
    "xl": 16,
    "xxl": 24,
    "xxxl": 36,
    "relative": {
      "xxxs": 2,
      "xxs": 3,
      "xs": 4,
      "small": 6,
      "medium": 8,
      "large": 12,
      "xl": 16,
      "xxl": 24,
      "xxxl": 36,
      "full": 800
    }
  },
  "time": {
    "short": 100,
    "medium": 250,
    "long": 500
  },
  "transforms": {
    "slideX": {
      "small": "translateX(0.5em)",
      "medium": "translateX(1em)",
      "large": "translateX(2em)"
    }
  }
};
    export default theme;
```

#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are updated
- [ ] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/ui/sd.config.ts [hunk 1]
```diff
@@ -40,5 +47,25 @@ StyleDictionary.extend({
...
+        'RNOpacity',
+        'RNTime',
+        'RNShadow',
+      ],
+      files: [
+        {
+          destination: 'dist/react-native/theme.ts',
+          format: 'RNFormat',
+          filter: (token) =>
+            !token.path.some((p) => ['hover', 'focus'].includes(p)),
```
Comment (joebuono) [inline packages/ui/sd.config.ts:66]: Just wondering, are there any other states we should exclude?

File: packages/ui/src/theme/platforms/react-native/matchers.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { DesignToken } from 'style-dictionary/types/DesignToken';
+
+const isRadius = (token: DesignToken): boolean =>
```
Comment (joebuono) [inline packages/ui/src/theme/platforms/react-native/matchers.ts:3]: nit: Can we alphabetize these functions in the same order that they're exported?

File: packages/ui/src/theme/tokens/types/designToken.ts [hunk 3]
```diff
@@ -43,6 +43,24 @@ export type WebDesignToken<ValueType = any> = {
   toString(): string;
 } & DesignToken<ValueType>;
 
+/**
+ * A fully setup design token ready to be used in mobile platform.
+ */
+export type ReactNativeDesignToken<ValueType = any> = {
```
Comment (joebuono) [inline packages/ui/src/theme/tokens/types/designToken.ts:49]: Can we be more specific than `ValueType = any`?

File: packages/ui/src/theme/platforms/react-native/transforms.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      * auto
+      * padding
+      * gap
+      * boxShadow
+      * remove any units (px, rem, em)
+*/
+
+export const ReactNativeTransforms: Record<string, Transform> = {
+  RNFontWeight: {
+    type: 'value',
```
Comment (joebuono) [inline packages/ui/src/theme/platforms/react-native/transforms.ts:26]: Just curious, why do we have to specify `type: 'value'` for each of these transforms?

File: packages/ui/src/theme/platforms/react-native/transforms.ts [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    matcher: isFontWeight,
+    transformer: (token: DesignToken) => {
+      const intValue = parseInt(token.value, 10);
```
Comment (joebuono) [inline packages/ui/src/theme/platforms/react-native/transforms.ts:29]: I see that some of these transforms use `parseInt`, `parseFloat`, and `Math.floor`. I'm wondering which of these values can actually be floats, or if we can just use `parseInt` for everything.

Comment (ioanabrooks) [reply packages/ui/src/theme/platforms/react-native/transforms.ts:29]: This is based on the current values in the corresponding design tokens. For example font sizes are in rems and are floats: https://github.com/aws-amplify/amplify-ui/blob/main/packages/ui/src/theme/tokens/fontSizes.ts

File: packages/ui/src/theme/platforms/react-native/transforms.ts (hunk 4 above)
Comment (ioanabrooks) [reply packages/ui/src/theme/platforms/react-native/transforms.ts:26]: https://github.com/aws-amplify/amplify-ui/blob/main/packages/ui/src/theme/tokens/types/designToken.ts#L20

**Pull request closed without merging**